In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [2]:
df = pd.read_csv('kc_house_data.csv')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [3]:
print(f"Le dataFrame comporte {df.shape[0]} lignes, pour {df.shape[1]} colonnes. ")

Le dataFrame comporte 21613 lignes, pour 21 colonnes. 


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [5]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [6]:
df['date'] = pd.to_datetime(df['date'])

In [7]:
df_sans_doublon = df.drop_duplicates('id')
df_sans_doublon.shape

(21436, 21)

In [8]:
Liste_maison_vendues_plusieurs_fois = df[df.duplicated('id')]
print(Liste_maison_vendues_plusieurs_fois['id'])

94       6021501535
314      4139480200
325      7520000520
346      3969300030
372      2231500030
            ...    
20181    7853400250
20613    2724049222
20670    8564860270
20780    6300000226
21581    7853420110
Name: id, Length: 177, dtype: int64


Il y a 177 maisons qui ont été vendues plusieurs fois.

In [9]:
df.loc[df['id']==6021501535]

id       date     price  bedrooms  bathrooms  sqft_living  \
93  6021501535 2014-07-25  430000.0         3        1.5         1580   
94  6021501535 2014-12-23  700000.0         3        1.5         1580   

    sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
93      5000     1.0           0     0  ...      8        1290            290   
94      5000     1.0           0     0  ...      8        1290            290   

    yr_built  yr_renovated  zipcode     lat     long  sqft_living15  \
93      1939             0    98117  47.687 -122.386           1570   
94      1939             0    98117  47.687 -122.386           1570   

    sqft_lot15  
93        4500  
94        4500  

[2 rows x 21 columns]

In [10]:
df.drop_duplicates('id').index[372]

377

In [11]:
df.iloc[377] == df.drop_duplicates('id').iloc[372]

id               True
date             True
price            True
bedrooms         True
bathrooms        True
sqft_living      True
sqft_lot         True
floors           True
waterfront       True
view             True
condition        True
grade            True
sqft_above       True
sqft_basement    True
yr_built         True
yr_renovated     True
zipcode          True
lat              True
long             True
sqft_living15    True
sqft_lot15       True
Name: 377, dtype: bool

test pour vérifier nos hypothèses.

In [12]:
duplicate_ids = df.loc[df['id'].isin(df['id'].value_counts()[df['id'].value_counts() > 1].index)]
duplicate_ids.head(6)

id       date      price  bedrooms  bathrooms  sqft_living  \
93   6021501535 2014-07-25   430000.0         3       1.50         1580   
94   6021501535 2014-12-23   700000.0         3       1.50         1580   
313  4139480200 2014-06-18  1384000.0         4       3.25         4290   
314  4139480200 2014-12-09  1400000.0         4       3.25         4290   
324  7520000520 2014-09-05   232000.0         2       1.00         1240   
325  7520000520 2015-03-11   240500.0         2       1.00         1240   

     sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
93       5000     1.0           0     0  ...      8        1290   
94       5000     1.0           0     0  ...      8        1290   
313     12103     1.0           0     3  ...     11        2690   
314     12103     1.0           0     3  ...     11        2690   
324     12092     1.0           0     0  ...      6         960   
325     12092     1.0           0     0  ...      6         960   

     sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
93             290      1939             0    98117  47.6870 -122.386   
94             290      1939             0    98117  47.6870 -122.386   
313           1600      1997             0    98006  47.5503 -122.102   
314           1600      1997             0    98006  47.5503 -122.102   
324            280      1922          1984    98146  47.4957 -122.352   
325            280      1922          1984    98146  47.4957 -122.352   

     sqft_living15  sqft_lot15  
93            1570        4500  
94            1570        4500  
313           3860       11244  
314           3860       11244  
324           1820        7460  
325           1820        7460  

[6 rows x 21 columns]

la liste exhaustive des ventes des maisons vendues plusieurs fois.

In [13]:
grouped_duplicates = duplicate_ids.groupby('id')

changes = grouped_duplicates.apply(lambda x: (x.iloc[0] != x).any())

changes = changes.drop('id', axis=1)
changes.head()

date  price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                           
1000102    True   True     False      False        False     False   False   
7200179    True   True     False      False        False     False   False   
109200390  True   True     False      False        False     False   False   
123039336  True   True     False      False        False     False   False   
251300110  True   True     False      False        False     False   False   

           waterfront   view  condition  grade  sqft_above  sqft_basement  \
id                                                                          
1000102         False  False      False  False       False          False   
7200179         False  False      False  False       False          False   
109200390       False  False      False  False       False          False   
123039336       False  False      False  False       False          False   
251300110       False  False      False  False       False          False   

           yr_built  yr_renovated  zipcode    lat   long  sqft_living15  \
id                                                                        
1000102       False         False    False  False  False          False   
7200179       False         False    False  False  False          False   
109200390     False         False    False  False  False          False   
123039336     False         False    False  False  False          False   
251300110     False         False    False  False  False          False   

           sqft_lot15  
id                     
1000102         False  
7200179         False  
109200390       False  
123039336       False  
251300110       False

Parmis les id des maisons vendues plusieurs fois, nous avons la liste des paramètres ayant changé ou pas d'une vente à l'autre.

In [14]:
changes.loc[(changes['price'] == False) ]

date  price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                            
1825069031  True  False     False      False        False     False   False   
6308000010  True  False     False      False        False     False   False   
8648900110  True  False     False      False        False     False   False   

            waterfront   view  condition  grade  sqft_above  sqft_basement  \
id                                                                           
1825069031       False  False      False  False       False          False   
6308000010       False  False      False  False       False          False   
8648900110       False  False      False  False       False          False   

            yr_built  yr_renovated  zipcode    lat   long  sqft_living15  \
id                                                                         
1825069031     False         False    False  False  False          False   
6308000010     False         False    False  False  False          False   
8648900110     False         False    False  False  False          False   

            sqft_lot15  
id                      
1825069031       False  
6308000010       False  
8648900110       False

Les id des maisons vendues plusieurs fois mais qu'aucun paramètre n'a changé, pas même le prix. Il faudra surement garder que la vente la plus récente en date.

In [15]:
for column in changes:
    print(column,changes[column].unique())

date [ True]
price [ True False]
bedrooms [False]
bathrooms [False]
sqft_living [False]
sqft_lot [False]
floors [False]
waterfront [False]
view [False]
condition [False]
grade [False]
sqft_above [False]
sqft_basement [False]
yr_built [False]
yr_renovated [False]
zipcode [False]
lat [False]
long [False]
sqft_living15 [False]
sqft_lot15 [False]


aucune feature à part le prix change d'une vente à l'autre, donc nous n'avons pas de justification pour un changement de prix à l'aide des données fournies. Nous allons donc simplement garder la dernière version en date.

In [16]:
def convert_date(date_str):
    date_dt = pd.to_datetime(date_str, format="%Y%m%dT%H%M%S")
    return date_dt.strftime("%Y/%m/%d")

In [20]:
duplicate_ids.groupby('id').apply(lambda x : x.idxmax())

id   date  price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id                                                                            
1000102      2496   2497   2497      2496       2496         2496      2496   
7200179     16814  16815  16815     16814      16814        16814     16814   
109200390   11433  11434  11434     11433      11433        11433     11433   
123039336   12417  12418  12418     12417      12417        12417     12417   
251300110    7792   7793   7793      7792       7792         7792      7792   
...           ...    ...    ...       ...        ...          ...       ...   
9407110710   2493   2494   2494      2493       2493         2493      2493   
9809000020   4922   4923   4923      4922       4922         4922      4922   
9828200460   6345   6346   6346      6345       6345         6345      6345   
9834200305  15199  15200  15200     15199      15199        15199     15199   
9834200885   1085   1086   1086      1085       1085         1085      1085   

            floors  waterfront   view  ...  grade  sqft_above  sqft_basement  \
id                                     ...                                     
1000102       2496        2496   2496  ...   2496        2496           2496   
7200179      16814       16814  16814  ...  16814       16814          16814   
109200390    11433       11433  11433  ...  11433       11433          11433   
123039336    12417       12417  12417  ...  12417       12417          12417   
251300110     7792        7792   7792  ...   7792        7792           7792   
...            ...         ...    ...  ...    ...         ...            ...   
9407110710    2493        2493   2493  ...   2493        2493           2493   
9809000020    4922        4922   4922  ...   4922        4922           4922   
9828200460    6345        6345   6345  ...   6345        6345           6345   
9834200305   15199       15199  15199  ...  15199       15199          15199   
9834200885    1085        1085   1085  ...   1085        1085           1085   

            yr_built  yr_renovated  zipcode    lat   long  sqft_living15  \
id                                                                         
1000102         2496          2496     2496   2496   2496           2496   
7200179        16814         16814    16814  16814  16814          16814   
109200390      11433         11433    11433  11433  11433          11433   
123039336      12417         12417    12417  12417  12417          12417   
251300110       7792          7792     7792   7792   7792           7792   
...              ...           ...      ...    ...    ...            ...   
9407110710      2493          2493     2493   2493   2493           2493   
9809000020      4922          4922     4922   4922   4922           4922   
9828200460      6345          6345     6345   6345   6345           6345   
9834200305     15199         15199    15199  15199  15199          15199   
9834200885      1085          1085     1085   1085   1085           1085   

            sqft_lot15  
id                      
1000102           2496  
7200179          16814  
109200390        11433  
123039336        12417  
251300110         7792  
...                ...  
9407110710        2493  
9809000020        4922  
9828200460        6345  
9834200305       15199  
9834200885        1085  

[176 rows x 21 columns]

à revoir 